linkerd/app/outbound/src/http/logical.rs

```diff
@@ -1,143 +1,270 @@
 use super::{retry, CanonicalDstHeader, Concrete, Logical};
-use crate::{stack_labels, Outbound};
+use crate::Outbound;
 use linkerd_app_core::{
-    classify, metrics, profiles,
-    proxy::{api_resolve::ConcreteAddr, http},
-    svc, Error, Infallible,
+    classify, metrics,
+    profiles::{self, Profile},
+    proxy::api_resolve::ConcreteAddr,
+    proxy::http,
+    svc, Error,
 };
-use tracing::debug_span;
+use linkerd_distribute as distribute;
+use std::sync::Arc;
+
+#[derive(Debug, thiserror::Error)]
+#[error("no route")]
+pub struct NoRoute;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct Params {
+    logical: Logical,
+    routes: Arc<[(profiles::http::RequestMatch, RouteParams)]>,
+    backends: distribute::Backends<Concrete>,
+}
 
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
-struct ProfileRoute {
+struct RouteParams {
     logical: Logical,
-    route: profiles::http::Route,
+    profile: profiles::http::Route,
+    distribution: Distribution,
 }
 
+type BackendCache<N, S> = distribute::BackendCache<Concrete, N, S>;
+type Distribution = distribute::Distribution<Concrete>;
+
+// === impl Outbound ===
+
 impl<N> Outbound<N> {
-    pub fn push_http_logical<NSvc>(self) -> Outbound<svc::ArcNewHttp<Logical>>
+    /// Builds a `NewService` that produces a router service for each logical
+    /// target.
+    ///
+    /// The router uses discovery information (provided on the target) to
+    /// support per-request routing over a set of concrete inner services.
+    /// Only available inner services are used for routing. When there are no
+    /// available backends, requests are failed with a [`svc::stack::LoadShedError`].
```

Comment:
nit, not a big deal: if we're gonna document the error types here, should we maybe also state that the `NoRoute` error is returned if there's no route that matches a request?